File: clients/csharp/NerveClient/NexusClient.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Sockets;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace Alenia.Nerve
{
    public class NexusClientConfig
    {
        public string SocketPath { get; set; } = "/tmp/nerve.sock";
        public string Host { get; set; } = "127.0.0.1";
        public int Port { get; set; } = 50505;
        public string AuthToken { get; set; } = "";
        public bool UseTcp { get; set; } = false;
    }

    public class NexusClient : IDisposable
    {
        private readonly NexusClientConfig _config;
        private string _clientId = string.Empty;
        private Socket? _socket;
        private Stream? _stream;
        private StreamReader? _reader;
        private StreamWriter? _writer;
        
        private bool _connected;
        private bool _closed = true;
        private readonly SemaphoreSlim _writeLock = new SemaphoreSlim(1, 1);
        private Thread? _receiveThread;

        private readonly Dictionary<string, List<Action<string, JObject>>> _handlers = new();
        private readonly object _handlersLock = new object();

        public NexusClient(NexusClientConfig config)
        {
            _config = config;
            
            // Auto force TCP on Windows if not already set, since UDS paths like /tmp/nerve.sock won't work easily
            if (Environment.OSVersion.Platform == PlatformID.Win32NT)
            {
                _config.UseTcp = true;
            }
        }

        public void Connect(string clientId)
        {
            _clientId = clientId;
            _closed = false;

            DoConnect();

            _receiveThread = new Thread(ReceiveLoop) { IsBackground = true };
            _receiveThread.Start();
        }

        private void DoConnect()
        {
            if (_closed) return;

            try
            {
                if (_config.UseTcp)
                {
                    _socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
                    _socket.Connect(_config.Host, _config.Port);
                }
                else
                {
                    _socket = new Socket(AddressFamily.Unix, SocketType.Stream, ProtocolType.Unspecified);
                    _socket.Connect(new UnixDomainSocketEndPoint(_config.SocketPath));
                }

                _stream = new NetworkStream(_socket);
                _reader = new StreamReader(_stream, Encoding.UTF8);
                _writer = new StreamWriter(_stream, Encoding.UTF8) { AutoFlush = true };

                _connected = true;
                Console.WriteLine($"[NERVE C#] Connected to hub as '{_clientId}'");

                var regMsg = new JObject
                {
                    ["type"] = "register",
                    ["id"] = _clientId
                };
                if (!string.IsNullOrEmpty(_config.AuthToken))
                {
                    regMsg["token"] = _config.AuthToken;
                }

                DoWrite(regMsg.ToString(Formatting.None) + "\n");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[NERVE C#] Connection failed: {ex.Message}");
                _connected = false;
            }
        }

        private void ReceiveLoop()
        {
            while (!_closed)
            {
                try
                {
                    if (_connected && _reader != null)
                    {
                        string? line = _reader.ReadLine();
                        if (line == null)
                        {
                            throw new IOException("Socket closed");
                        }
                        ProcessLine(line);
                    }
                    else
                    {
                        Thread.Sleep(2000);
                        DoConnect();
                    }
                }
                catch (Exception ex)
                {
                    if (!_closed)
                    {
                        Console.WriteLine($"[NERVE C#] Disconnected: {ex.Message}");
                        _connected = false;
                        Thread.Sleep(2000);
                        DoConnect();
                    }
                }
            }
        }

        private void ProcessLine(string line)
        {
            try
            {
                var parsed = JObject.Parse(line);
                var type = parsed["type"]?.ToString() ?? "";

                if (type == "ping")
                {
                    var pong = new JObject { ["type"] = "pong" };
                    DoWrite(pong.ToString(Formatting.None) + "\n");
                    return;
                }

                lock (_handlersLock)
                {
                    if (_handlers.TryGetValue(type, out var list))
                    {
                        foreach (var cb in list)
                        {
                            // Invoke handlers asynchronously to prevent blocking the read loop
                            Task.Run(() => cb(type, parsed));
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[NERVE C#] JSON parsing error: {ex.Message}");
            }
        }

        private void DoWrite(string msg)
        {
            if (!_connected || _writer == null) return;

            _writeLock.Wait();
            try
            {
                _writer.Write(msg);
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[NERVE C#] Write error: {ex.Message}");
                _connected = false;
            }
            finally
            {
                _writeLock.Release();
            }
        }

        public void Send(string to, JObject payload)
        {
            if (!_connected || _closed) return;
            var msg = new JObject
            {
                ["type"] = "send",
                ["to"] = to,
                ["payload"] = payload
            };
            DoWrite(msg.ToString(Formatting.None) + "\n");
        }

        public void Broadcast(JObject payload)
        {
            if (!_connected || _closed) return;
            var msg = new JObject
            {
                ["type"] = "broadcast",
                ["payload"] = payload
            };
            DoWrite(msg.ToString(Formatting.None) + "\n");
        }

        public void Listen(string eventType, Action<string, JObject> callback)
        {
            lock (_handlersLock)
            {
                if (!_handlers.ContainsKey(eventType))
                {
                    _handlers[eventType] = new List<Action<string, JObject>>();
                }
                _handlers[eventType].Add(callback);
            }
        }

        public void Disconnect()
        {
            if (_closed) return;
            _closed = true;
            _connected = false;

            try
            {
                _socket?.Close();
            }
            catch { }
        }

        public void Dispose()
        {
            Disconnect();
        }
    }
}

File: clients/cpp/src/nexus_client.cpp
#include "nerve/nexus_client.hpp"
#include <iostream>

#ifdef _WIN32
#define NERVE_USE_TCP 1
#else
#define NERVE_USE_TCP 0
#endif

namespace nerve {

NexusClient::NexusClient(Config cfg) : config_(std::move(cfg)) {
#if NERVE_USE_TCP
    config_.use_tcp = true; // Force TCP on Windows
#endif
}

NexusClient::~NexusClient() {
    disconnect();
}

void NexusClient::connect(const std::string& client_id) {
    client_id_ = client_id;
    closed_ = false;

    do_connect();

    io_thread_ = std::thread([this]() {
        while (!closed_) {
            try {
                io_context_.run();
                if (!closed_) {
                    std::this_thread::sleep_for(std::chrono::seconds(2));
                    io_context_.restart();
                    do_connect();
                }
            } catch (const std::exception& e) {
                std::cerr << "[NERVE C++] IO Error: " << e.what() << std::endl;
                if (!closed_) {
                    std::this_thread::sleep_for(std::chrono::seconds(2));
                    io_context_.restart();
                    do_connect();
                }
            }
        }
    });
}

void NexusClient::do_connect() {
    if (closed_) return;

    socket_ = std::make_unique<asio::generic::stream_protocol::socket>(io_context_);

    std::error_code ec;
    if (config_.use_tcp) {
        asio::ip::tcp::endpoint endpoint(asio::ip::make_address(config_.host), config_.port);
        socket_->connect(asio::generic::stream_protocol::endpoint(endpoint.protocol(), endpoint.data(), endpoint.size()), ec);
    } else {
        asio::local::stream_protocol::endpoint endpoint(config_.socket_path);
        socket_->connect(asio::generic::stream_protocol::endpoint(endpoint.protocol(), endpoint.data(), endpoint.size()), ec);
    }

    if (ec) {
        std::cerr << "[NERVE C++] Connection failed: " << ec.message() << std::endl;
        return;
    }

    connected_ = true;
    std::cout << "[NERVE C++] Connected to hub as '" << client_id_ << "'" << std::endl;

    json reg;
    reg["type"] = "register";
    reg["id"] = client_id_;
    if (!config_.auth_token.empty()) {
        reg["token"] = config_.auth_token;
    }

    do_write(reg.dump() + "\n");
    do_read();
}

void NexusClient::disconnect() {
    if (closed_) return;
    closed_ = true;
    connected_ = false;

    if (socket_) {
        std::error_code ec;
        socket_->close(ec);
    }
    io_context_.stop();

    if (io_thread_.joinable()) {
        io_thread_.join();
    }
}

void NexusClient::send(const std::string& to, const json& payload) {
    if (!connected_ || closed_) return;
    json msg;
    msg["type"] = "send";
    msg["to"] = to;
    msg["payload"] = payload;
    do_write(msg.dump() + "\n");
}

void NexusClient::broadcast(const json& payload) {
    if (!connected_ || closed_) return;
    json msg;
    msg["type"] = "broadcast";
    msg["payload"] = payload;
    do_write(msg.dump() + "\n");
}

void NexusClient::listen(const std::string& event_type, MessageCallback callback) {
    std::lock_guard<std::mutex> lock(handlers_mutex_);
    handlers_[event_type].push_back(std::move(callback));
}

void NexusClient::do_write(const std::string& msg) {
    if (!socket_ || !socket_->is_open()) return;
    
    std::lock_guard<std::mutex> lock(write_mutex_);
    std::error_code ec;
    asio::write(*socket_, asio::buffer(msg), ec);
    if (ec) {
        std::cerr << "[NERVE C++] Write error: " << ec.message() << std::endl;
        connected_ = false;
        socket_->close(ec);
    }
}

void NexusClient::do_read() {
    if (!socket_ || !socket_->is_open()) return;

    asio::async_read_until(*socket_, read_buffer_, '\n',
        [this](std::error_code ec, std::size_t length) {
            if (!ec) {
                std::istream is(&read_buffer_);
                std::string line;
                std::getline(is, line);
                if (!line.empty() && line.back() == '\r') {
                    line.pop_back();
                }
                
                process_line(line);
                do_read();
            } else {
                std::cerr << "[NERVE C++] Read error: " << ec.message() << std::endl;
                connected_ = false;
                std::error_code ignore_ec;
                socket_->close(ignore_ec);
            }
        });
}

void NexusClient::process_line(const std::string& line) {
    try {
        json parsed = json::parse(line);
        std::string type = parsed.value("type", "");

        if (type == "ping") {
            json pong;
            pong["type"] = "pong";
            do_write(pong.dump() + "\n");
            return;
        }

        std::lock_guard<std::mutex> lock(handlers_mutex_);
        if (handlers_.count(type)) {
            for (auto& cb : handlers_[type]) {
                cb(type, parsed);
            }
        }
    } catch (const std::exception& e) {
        std::cerr << "[NERVE C++] JSON parsing error: " << e.what() << std::endl;
    }
}

} // namespace nerve
